src/engine-actions/get-payment-preimage.js

```diff
@@ -27,62 +20,82 @@ const POLLING_INTERVAL = 10000
-    return { permanentError: `No payment with hash '${paymentHash}' is in-flight or complete` }
-  }
+      // TODO: look into how gRPC wants `bytes` in JS
+      const paymentHashBuf = Buffer.from(paymentHash, 'base64')
 
-  if (status === PAYMENT_STATUSES.COMPLETED) {
-    return { paymentPreimage: await getCompletedPreimage(paymentHash, { logger, client }) }
-  }
+      this.logger.debug('Attempting to track payment for paymentHash:', { paymentHash })
 
-  if (status === PAYMENT_STATUSES.IN_FLIGHT) {
-    // if the payment is still in-flight, we'll wait and try again later
-    await delay(POLLING_INTERVAL)
-    return getPaymentPreimage.call(this, paymentHash)
-  }
+      const stream = await trackPayment(paymentHashBuf, { client: this.client })
 
-  throw new Error(`Unknown payment status ${status} for payment with hash '${paymentHash}'`)
-}
+      const errorListener = (err) => {
+        // TODO: handle gRPC NOT FOUND errors
+        //  It may be best to handle it as a `permanentError`, so we don't have to
+        //  rely on checking NOT FOUND in downstream logic
+        this.logger.error(`Error from trackPayment stream`)
+        return finish(err)
+      }
 
-/**
- * Retrieve the preimage for a completed payment specified by the payment
- * hash.
- * @todo LND has a known issue preventing payments that are settled after a restart
- *       showing up in this list, which will mistakenly throw an error.
- *       {@link https://github.com/lightningnetwork/lnd/issues/2032}
- * @param  {string} paymentHash - Base64 encoded payment hash
- * @param  {Object} options
- * @param  {LndClient} options.client
- * @returns {string} Base64 encoded payment preimage associated with the hash
- * @throws {Error} If payment is not completed
- */
-async function getCompletedPreimage (paymentHash, { logger, client }) {
-  const { payments = [] } = await listPayments({ client })
+      const endListener = () => {
+        this.logger.error('LND closed trackPayment stream before returning our value')
+        return finish(new Error('LND closed stream to track payment'))
+      }
 
-  logger.debug(`Found ${payments.length} completed payments`)
+      const dataListener = (response) => {
+        const { state, preimage } = response
 
-  const paymentHashBuf = Buffer.from(paymentHash, 'base64')
-  const payment = payments.find(({ paymentHash }) => {
-    // `paymentHash` returned from ListPayments {@link https://api.lightning.community/#payment}
-    // is a hex string. We need to compare to our payment hash which is a base64 string,
-    // so we convert both to buffers to do the comparison.
-    return paymentHashBuf.equals(Buffer.from(paymentHash, 'hex'))
-  })
+        if (state === PAYMENT_STATUSES.IN_FLIGHT) {
+          // trackPayment will continue to stream updates of the payment state, so
+          // no action is required on our end
+          this.logger.debug('Payment is In-Flight for paymentHash:', { paymentHash })
```

Comment:
you're missing a `return` here - as is it will fall through to the `throw`

Reply:
Good catch.